oma-pm/src/apt.rs

```diff
@@ -908,31 +907,42 @@ impl OmaApt {
 
                 let size = cand.installed_size();
 
-                let mut entry = InstallEntryBuilder::default();
-                entry.name(pkg.fullname(true).to_string());
-                entry.new_version(version.to_string());
-                entry.new_size(size);
-                entry.pkg_urls(uri);
-                entry.arch(cand.arch().to_string());
-                entry.download_size(cand.size());
-                entry.op(InstallOperation::Install);
-                entry.automatic(!self.select_pkgs.contains(&pkg.fullname(true)));
-
-                if not_local_source {
-                    if sha256
+                if not_local_source
+                    && sha256
                         .as_ref()
                         .or(md5.as_ref())
                         .or(sha512.as_ref())
                         .is_none()
-                    {
-                        return Err(OmaAptError::PkgNoChecksum(pkg.to_string()));
-                    }
-                    entry.sha256(sha256);
-                    entry.md5(md5);
-                    entry.sha512(sha512);
+                {
+                    return Err(OmaAptError::PkgNoChecksum(pkg.to_string()));
                 }
 
-                let entry = entry.build()?;
+                let entry = if not_local_source {
+                    InstallEntry::builder()
+                        .name(pkg.fullname(true).to_string())
+                        .new_version(version.to_string())
+                        .new_size(size)
+                        .pkg_urls(uri)
+                        .arch(cand.arch().to_string())
+                        .download_size(cand.size())
+                        .op(InstallOperation::Install)
+                        .automatic(!self.select_pkgs.contains(&pkg.fullname(true)))
+                        .maybe_md5(md5)
+                        .maybe_sha256(sha256)
+                        .maybe_sha512(sha512)
+                        .build()
+                } else {
+                    InstallEntry::builder()
+                        .name(pkg.fullname(true).to_string())
+                        .new_version(version.to_string())
+                        .new_size(size)
+                        .pkg_urls(uri)
+                        .arch(cand.arch().to_string())
+                        .download_size(cand.size())
+                        .op(InstallOperation::Install)
+                        .automatic(!self.select_pkgs.contains(&pkg.fullname(true)))
+                        .build()
```

Comment:
Hi @eatradish , I found this commit throught the dependency graph in Github.

I recommend you to use some of the [patterns for conditional building](https://elastio.github.io/bon/guide/patterns/conditional-building#shared-partial-builder) to avoid repetition. For example, you can avoid extract the common "shared partial builder" here like this:

```rust
let entry = InstallEntry::builder()
    .name(pkg.fullname(true).to_string())
    .new_version(version.to_string())
    .new_size(size)
    .pkg_urls(uri)
    .arch(cand.arch().to_string())
    .download_size(cand.size())
    .op(InstallOperation::Install)
    .automatic(!self.select_pkgs.contains(&pkg.fullname(true)));

let entry = if not_local_source {
    entry
        .maybe_md5(md5)
        .maybe_sha256(sha256)
        .maybe_sha512(sha512)
        .build()
} else {
    entry.build()
};
```

Alternatively, if you are using `maybe_` methods (ones that take `Option` input) inside of the `if` branches, then you could extract them with the ["shared total builder"](https://elastio.github.io/bon/guide/patterns/conditional-building#shared-total-builder) like this then:

```rust
if not_local_source {
    md5 = None;
    sha256 = None;
    sha512 = None;
}

let entry = InstallEntry::builder()
    .name(pkg.fullname(true).to_string())
    .new_version(version.to_string())
    .new_size(size)
    .pkg_urls(uri)
    .arch(cand.arch().to_string())
    .download_size(cand.size())
    .op(InstallOperation::Install)
    .automatic(!self.select_pkgs.contains(&pkg.fullname(true)))
    .maybe_md5(md5)
    .maybe_sha256(sha256)
    .maybe_sha512(sha512)
    .build();
```

Note that I'm not familiar with the business logic of this code. I'm not sure why you'd not want to set the hashes for `non_local_source` especially if they are calculated, so this code looks a bit suspicios. Could it just always uncoditionally set `maybe_{hash}` members with their respective `Option<T>` values?

Reply:
Hi, thanks for the suggestion, I've come up with #80 based on your suggestion. If you're interested, please come over for another review!